pydocstringformatter/_formatting/base.py

```diff
@@ -202,3 +205,61 @@ def treat_summary(
 
     def treat_description(self, description: str, indent_length: int) -> str:
         return description
+
+
+def _pairwise(iterator: Iterator[_T]) -> Iterator[tuple[_T, _T]]:
+    """Create an iterator over pairs of successive elements."""
+    first, second = tee(iterator)
+    next(second)
+    return zip(first, second)
+
+
+class NumpydocSectionFormatter(StringAndQuotesFormatter, metaclass=abc.ABCMeta):
+    """Base class for formatters working on numpydoc sections."""
+
+    style = ["numpydoc"]
+
+    @abc.abstractmethod
+    def treat_sections(self, sections: list[list[str]]) -> list[list[str]]:
+        """Process the individual numpydoc sections."""
+
+    def treat_string(
+        self,
+        tokeninfo: tokenize.TokenInfo,
+        indent_length: int,
+        quotes: str,
+        quotes_length: Literal[1, 3],
+    ) -> str:
+        """Split numpydoc sections, pass them for processing, then rejoin them."""
+        # Split sections
+        lines = tokeninfo.string[quotes_length:-quotes_length].split("\n")
+        if lines[-1].isspace():
+            last_line = lines[-1]
+            lines[-1] = ""
+        else:
+            last_line = ""
+
+        section_heading_lines = [
+            index
+            for index, line in enumerate(lines)
+            if "-" in line and all(char in " \t-" for char in line)
+        ]
+        section_starts = (
+            [0] + [index - 1 for index in section_heading_lines] + [len(lines)]
+        )
+
+        sections = [
+            lines[curr_section_start:next_section_start]
+            for curr_section_start, next_section_start in _pairwise(
+                iter(section_starts)
+            )
+        ]
+        # Process sections
+        new_sections = self.treat_sections(sections)
+
+        # Rejoin sections
+        new_lines = [line for section in new_sections for line in section]
+        if len(last_line) > 0:
+            new_lines[-1] = last_line
```

Comment:
Should this be appended or overwritten?

Reply:
The sections are *usually* separated by a blank line, but the last one needs spaces so the closing quotes are indented properly.  I should probably write a formatter for the first one and check which I need to do for the second (if it's a blank line, I replace it with the line of spaces; if not, I append the line of spaces).

Reply:
Can't you use `indent_length` for this? It is in the arguments to this function specifically for these situations.

Reply:
```suggestion
            if new_lines[-1] and new_lines[-1].isspace():
                new_lines[-1] = last_line
            else:
                new_lines.append(last_line)
```

Reply:
> Can't you use `indent_length` for this? It is in the arguments to this function specifically for these situations.

I'd forgotten about that.  Yes.